bundles/org.openhab.automation.jsscripting/pom.xml

```diff
@@ -0,0 +1,140 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright (c) 2010-2020 Contributors to the openHAB project
+
+  See the NOTICE file(s) distributed with this work for additional
+  information.
+
+  This program and the accompanying materials are made available under the
+  terms of the Eclipse Public License 2.0 which is available at
+  http://www.eclipse.org/legal/epl-2.0
+
+  SPDX-License-Identifier: EPL-2.0
+-->
+
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.0.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>org.openhab.automation.jsscripting</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: Automation :: JSScripting</name>
+
+  <properties>
+    <bnd.importpackage>
+      !sun.misc.*,
+      !sun.reflect.*,
+      !com.sun.management.*,
+      !jdk.internal.reflect.*,
+      !jdk.vm.ci.services
+    </bnd.importpackage>
+    <graal.version>20.1.0</graal.version>
+    <asm.version>6.2.1</asm.version>
+    <oh.version>${project.version}</oh.version>
+  </properties>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-dependency-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>embed-dependencies</id>
```

Comment:
Did you look into embedding this content with BND?

Reply:
I have not, but I will do...

Reply:
The reason that this is in here is because Graal uses ServiceLoaders to load its language definitions, and has one per language per jar. This means that for each jar there is a single `META-INF/services/com.oracle.truffle.api.TruffleLanguage$Provider` file which is used to inform the ServiceLoader. The problem we have is that we need two languages (javascript & tregex), but when the multiple jars are all expanded and added into this bundle, we end up with a single one of the above files (as they are present at the same path in the jar). The way that I have got around this is to provide my own version of this file, with both languages in it, and exclude the ones coming from the Graal jars (this declaration in the pom.xml just overrides the inherited one and instructs it to exclude this file).

It seems that there are a few options here:
- Do what I've done so far
- Attempt to merge the files from the distinct jars (I tried this, it was a lot of complexity wrestling with Maven)
- Bundle the jars into META-INF/lib rather than repack them (I'm aware that this isn't recommended for openHAB)
- Attempt to put my own file over the top of whatever gets included by inserted it at a later build phase. Maybe this is what you mean by using BND?

Reply:
Sounds like something non-trivial to rework. So let's stick with what works. :wink: